staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -420,14 +429,40 @@ func (s *GenericAPIServer) installAPIResources(apiPrefix string, apiGroupInfo *A
 		apiGroupVersion.OpenAPIModels = openAPIModels
 		apiGroupVersion.MaxRequestBodyBytes = s.maxRequestBodyBytes
 
-		if err := apiGroupVersion.InstallREST(s.Handler.GoRestfulContainer); err != nil {
+		r, err := apiGroupVersion.InstallREST(s.Handler.GoRestfulContainer)
+		if err != nil {
 			return fmt.Errorf("unable to setup API %v: %v", apiGroupInfo, err)
 		}
+		resourceInfos = append(resourceInfos, r...)
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.StorageVersionAPI) {
+		// API installation happens before we start listening on the handlers,
+		// therefore it is safe to register ResourceInfos here. The handler will block
+		// write requests until the targeting resources' storage version is up-to-date.
+		s.StorageVersion.AddResourceInfo(dedupResourceInfos(resourceInfos)...)
 	}
 
 	return nil
 }
 
+// dedupResourceInfos dedups ResourceInfos with the same underlying storage.
+// ResourceInfos from different Groups may share the same underlying storage, e.g.
+// networking.k8s.io ingresses and extensions ingresses. The StorageVersion manager
+// only needs to update one StorageVersion for the equivalent Groups.
+func dedupResourceInfos(infos []*storageversion.ResourceInfo) []*storageversion.ResourceInfo {
```

Comment:
Do we still need this function? I remember we used the EquivalentResourceMapper to dedup.

 This function is problematic. If two resources are stored at different path in the storage, but share the same encoding version, they should be deduplicated.

Reply:
Switched to use the EquivalentResourceMapper to dedup.